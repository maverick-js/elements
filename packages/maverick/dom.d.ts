export * from './dist/prod/dom.js';
