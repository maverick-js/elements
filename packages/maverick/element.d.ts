export * from './dist/prod/element.js';
