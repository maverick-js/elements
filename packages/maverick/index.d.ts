export * from './dist/prod/runtime.js';
