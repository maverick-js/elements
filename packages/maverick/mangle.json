{
  "_$state": "h",
  "_addHooks": "t",
  "_attach": "x",
  "_attachAttrs": "q",
  "_attachCallbacks": "m",
  "_attachScope": "f",
  "_attachStyles": "r",
  "_attached": "Q",
  "_attrs": "b",
  "_callbacks": "J",
  "_component": "e",
  "_connect": "y",
  "_connectCallbacks": "j",
  "_connectId": "E",
  "_connectScope": "g",
  "_context": "F",
  "_converter": "B",
  "_currentRenderer": "X",
  "_currentRenderer2": "Y",
  "_currentValue": "M",
  "_currentValue2": "N",
  "_defaultValue": "V",
  "_descriptors": "A",
  "_destroy": "p",
  "_destroyCallbacks": "n",
  "_destroyed": "o",
  "_detach": "s",
  "_disconnect": "z",
  "_el": "a",
  "_globalName": "W",
  "_init": "S",
  "_owner": "G",
  "_payload": "R",
  "_prop": "C",
  "_props": "i",
  "_provide": "T",
  "_refChangeId": "K",
  "_result": "I",
  "_scope": "d",
  "_self": "U",
  "_setAttr": "u",
  "_setStyle": "v",
  "_setup": "w",
  "_setupCallbacks": "l",
  "_source": "L",
  "_state": "k",
  "_status": "D",
  "_store": "H",
  "_styles": "c",
  "_threadCount": "O",
  "_tokens": "Z",
  "_updatedFibers": "P",
  "_domCallbacks": "_",
  "_prevProps": "aa",
  "_onDOMEvent": "ba",
  "_forwardedRef": "da",
  "_listening": "ca",
  "_destroyId": "ea"
}