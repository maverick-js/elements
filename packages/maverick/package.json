{
  "name": "maverick.js",
  "description": "Build functional, reactive, type-safe, lightweight, and fast UI libraries.",
  "license": "MIT",
  "version": "0.32.3",
  "type": "module",
  "types": "dist/types/runtime/index.d.ts",
  "main": "dist/server/runtime.cjs",
  "module": "dist/prod/runtime.js",
  "sideEffects": false,
  "files": [
    "dist/",
    "*.d.ts"
  ],
  "engines": {
    "node": ">=16"
  },
  "contributors": [
    "Rahim Alwer <rahim.alwer@gmail.com>"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/maverick-js/maverick.git",
    "directory": "packages/maverick"
  },
  "bugs": {
    "url": "https://github.com/maverick-js/maverick/issues"
  },
  "scripts": {
    "dev": "pnpm run build --watch",
    "build": "rimraf dist && tsup && pnpm types",
    "types": "tsc -p tsconfig.build.json",
    "test": "pnpm test:client && pnpm test:server",
    "format": "prettier src --write --loglevel warn",
    "test:client": "vitest --run",
    "test:client:watch": "vitest --watch",
    "test:server": "SERVER=true vitest --run",
    "test:server:watch": "SERVER=true vitest --watch"
  },
  "dependencies": {
    "@maverick-js/signals": "^5.9.1",
    "type-fest": "^2.17.0"
  },
  "devDependencies": {
    "@edge-runtime/vm": "^2.0.0",
    "@maverick-js/compiler": "workspace:*",
    "@types/node": "^18.0.3",
    "@types/react": "^18.0.24",
    "@types/react-dom": "^18.0.8",
    "benchmark": "^2.1.4",
    "jsdom": "^20.0.0",
    "node-fetch": "^3.2.10",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rimraf": "^3.0.2",
    "tsup": "^6.2.3",
    "typescript": "^4.9.0",
    "vite": "^3.2.3",
    "vitest": "^0.25.1"
  },
  "exports": {
    ".": {
      "import": {
        "test": "./dist/dev/runtime.js",
        "node": "./dist/server/runtime.js",
        "deno": "./dist/server/runtime.js",
        "bun": "./dist/server/runtime.js",
        "development": "./dist/dev/runtime.js",
        "default": "./dist/prod/runtime.js"
      },
      "require": "./dist/server/runtime.cjs"
    },
    "./dom": {
      "import": {
        "test": "./dist/dev/dom.js",
        "node": "./dist/server/dom.js",
        "deno": "./dist/server/dom.js",
        "bun": "./dist/server/dom.js",
        "development": "./dist/dev/dom.js",
        "default": "./dist/prod/dom.js"
      },
      "require": "./dist/server/dom.cjs"
    },
    "./ssr": {
      "import": {
        "test": "./dist/dev/ssr.js",
        "node": "./dist/server/ssr.js",
        "deno": "./dist/server/ssr.js",
        "bun": "./dist/server/ssr.js",
        "development": "./dist/dev/ssr.js",
        "default": "./dist/prod/ssr.js"
      },
      "require": "./dist/server/ssr.cjs"
    },
    "./element": {
      "import": {
        "test": "./dist/dev/element.js",
        "node": "./dist/server/element.js",
        "deno": "./dist/server/element.js",
        "bun": "./dist/server/element.js",
        "development": "./dist/dev/element.js",
        "default": "./dist/prod/element.js"
      },
      "require": "./dist/server/element.cjs"
    },
    "./react": {
      "import": {
        "test": "./dist/dev/react.js",
        "node": "./dist/server/react.js",
        "deno": "./dist/server/react.js",
        "bun": "./dist/server/react.js",
        "development": "./dist/dev/react.js",
        "default": "./dist/prod/react.js"
      },
      "require": "./dist/server/react.cjs"
    },
    "./std": {
      "import": {
        "test": "./dist/dev/std.js",
        "node": "./dist/server/std.js",
        "deno": "./dist/server/std.js",
        "bun": "./dist/server/std.js",
        "development": "./dist/dev/std.js",
        "default": "./dist/prod/std.js"
      },
      "require": "./dist/server/std.cjs"
    },
    "./package.json": "./package.json"
  },
  "publishConfig": {
    "access": "public"
  }
}
