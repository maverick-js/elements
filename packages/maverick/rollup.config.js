import { nodeResolve } from '@rollup/plugin-node-resolve';
import { build } from 'esbuild';
import { globbySync } from 'globby';
import fs from 'node:fs/promises';
import { defineConfig } from 'rollup';
import dts from 'rollup-plugin-dts';
import esbuildPlugin from 'rollup-plugin-esbuild';

/** @type {Record<string, string | false>} */
const MANGLE_CACHE = await buildMangleCache();

export default defineConfig([
  define({ type: 'dev' }),
  define({ type: 'prod' }),
  define({ type: 'server' }),
  define({ type: 'rsc-dev' }),
  define({ type: 'rsc-prod' }),
  defineTypes(),
]);

/** @returns {import('rollup').RollupOptions} */
function defineTypes() {
  return {
    input: {
      index: 'types/core/index.d.ts',
      element: 'types/element/index.d.ts',
      react: 'types/react/index.d.ts',
      std: 'types/std/index.d.ts',
    },
    output: {
      dir: '.',
      chunkFileNames: 'dist/types/maverick-[hash].d.ts',
      compact: false,
      minifyInternalExports: false,
    },
    external: ['react'],
    plugins: [
      dts({ respectExternal: true }),
      {
        name: 'cleanup',
        async closeBundle() {
          await fs.rm('types', { recursive: true });
        },
      },
    ],
  };
}

/**
 * @typedef {{
 *   type: 'dev' | 'prod' | 'server' | 'rsc-dev' | 'rsc-prod';
 * }} BundleOptions
 */

/**
 * @param {BundleOptions}
 * @returns {import('rollup').RollupOptions}
 */
function define({ type = 'dev' }) {
  const isDev = type === 'dev' || type === 'rsc-dev',
    isRSC = type.startsWith('rsc');

  return {
    input: isRSC
      ? { [type.replace('rsc-', '')]: 'src/react/rsc.ts' }
      : {
          index: 'src/core/index.ts',
          element: 'src/element/index.ts',
          react: 'src/react/index.ts',
          std: 'src/std/index.ts',
        },
    treeshake: true,
    preserveEntrySignatures: 'allow-extension',
    external: ['react'],
    output: {
      format: 'esm',
      dir: isRSC ? `dist/rsc` : `dist/${type}`,
      chunkFileNames: `chunks/maverick-[hash].js`,
      compact: false,
      minifyInternalExports: false,
    },
    plugins: [
      nodeResolve({
        exportConditions: isDev
          ? ['development', 'production', 'default']
          : ['production', 'default'],
      }),
      esbuildPlugin({
        target: type === 'server' ? 'node18' : 'esnext',
        platform: type === 'server' ? 'node' : 'browser',
        tsconfig: 'tsconfig.build.json',
        minify: false,
        mangleProps: !isDev ? /^_/ : undefined,
        reserveProps: !isDev ? /^__/ : undefined,
        mangleCache: !isDev ? MANGLE_CACHE : undefined,
        banner: isRSC ? 'import { IS_SERVER } from "@virtual/env";\n' : '',
        define: {
          __DEV__: isDev ? 'true' : 'false',
          __SERVER__: isRSC ? 'IS_SERVER' : type === 'server' ? 'true' : 'false',
          __TEST__: 'false',
        },
      }),
      {
        name: 'env',
        resolveId(id) {
          if (id === '@virtual/env') return id;
        },
        load(id) {
          if (id === '@virtual/env') {
            return `export const IS_SERVER = typeof document === 'undefined';`;
          }
        },
      },
    ],
  };
}

async function buildMangleCache() {
  let cache = JSON.parse(await fs.readFile('mangle.json', 'utf-8'));

  const result = await build({
    entryPoints: globbySync('src/**'),
    target: 'esnext',
    bundle: true,
    minify: false,
    mangleProps: /^_/,
    reserveProps: /^__/,
    mangleCache: cache,
    write: false,
    outdir: 'dist-esbuild',
  });

  cache = {
    ...cache,
    ...result.mangleCache,
  };

  await fs.writeFile('mangle.json', JSON.stringify(cache, null, 2));

  return cache;
}
