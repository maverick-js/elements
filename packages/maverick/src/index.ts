export * from './core/component';
export * from './core/context';
export * from './core/controller';
export * from './core/decorators';
export * from './core/instance';
export * from './core/signals';
export * from './core/state';
export * from './core/types';
