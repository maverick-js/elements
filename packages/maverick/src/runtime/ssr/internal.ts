import { type AnyElementDefinition, createServerElement } from '../../element';
import { createElementInstance } from '../../element/instance';
import { getElementInstance } from '../../element/internal';
import { escape } from '../../std/html';
import { unwrapDeep } from '../../std/observable';
import { trimTrailingSemicolon } from '../../std/string';
import { isFunction, isNull, isString } from '../../std/unit';
import { injectHTML, resolve, SSR_TEMPLATE } from './render';

/** @internal */
export function $$_ssr(template: string[], ...parts: unknown[]) {
  let result = '';

  for (let i = 0; i < template.length; i++) {
    result += template[i];
    result += resolve(escape(parts[i]));
  }

  return { [SSR_TEMPLATE]: result };
}

/** @internal */
export function $$_host_element(spreads?: Record<string, unknown>[]) {
  const host = getElementInstance()!.host.el!;

  if (spreads && spreads.length > 0) {
    const spread = $$_merge_spreads(spreads);

    for (const [key, value] of spread.attributes) {
      host.setAttribute(key, value);
    }

    for (const token of spread.classList) {
      host.classList.add(token);
    }

    for (const [key, value] of spread.styles) {
      host.style.setProperty(key, value);
    }
  }

  return '';
}

/** @internal */
export function $$_custom_element(
  definition: AnyElementDefinition,
  props?: Record<string, any>,
  spreads?: Record<string, unknown>[],
) {
  const host = new (createServerElement(definition))();

  if (spreads && spreads.length > 0) {
    const spread = $$_merge_spreads(spreads);

    for (const [key, value] of spread.attributes) {
      host.attributes.setAttribute(key, value);
    }

    for (const token of spread.classList) {
      host.classList.add(token);
    }

    for (const [key, value] of spread.styles) {
      host.style.setProperty(key, value);
    }
  }

  const hasInnerHTML = !!props?.innerHTML,
    innerHTML = hasInnerHTML ? resolve(props.innerHTML) : null,
    children = hasInnerHTML ? innerHTML! : resolve(props?.$children),
    hasChildElements = children.replace(/<!(.*?)>/g, '').length > 0,
    instance = createElementInstance(definition, {
      props,
      children: () => hasChildElements,
    });

  host.attachComponent(instance);

  return {
    [SSR_TEMPLATE]: `<${definition.tagName}${host.attributes}>${
      hasInnerHTML ? children : host.render() + children
    }</${definition.tagName}>`,
  };
}

/** @internal */
export function $$_attr(name: string, value: unknown) {
  const attrValue = resolveAtrr(value);
  return isString(attrValue) ? ` ${name}="${escape(attrValue, true)}"` : '';
}

function resolveAtrr(value: unknown) {
  if (isFunction(value)) return resolveAtrr(value());
  if (!value && value !== '' && value !== 0) {
    return null;
  } else {
    return value + '';
  }
}

/** @internal */
export function $$_classes(base: unknown, tokens: Record<string, unknown>) {
  let baseValue = unwrapDeep(base),
    result = isString(baseValue) ? baseValue : '';

  if (Object.keys(tokens).length > 0) {
    const classList = new Set<string>();
    parseClassAttr(classList, result);
    resolveClasses(classList, tokens);
    result = Array.from(classList).join(' ');
  }

  result = result.trim();
  return result.length ? ` class="${escape(result, true)}"` : '';
}

const classSplitRE = /\s+/;
export function parseClassAttr(tokens: Set<string>, attrValue: string) {
  const classes = attrValue.trim().split(classSplitRE);
  for (const token of classes) tokens.add(token);
}

function resolveClasses(classList: Set<string>, tokens: Record<string, unknown>) {
  for (const name of Object.keys(tokens)) {
    if (unwrapDeep(tokens[name])) {
      classList.add(name);
    } else {
      classList.delete(name);
    }
  }
}

/** @internal */
export function $$_styles(base: unknown, tokens: Record<string, unknown>) {
  let baseValue = unwrapDeep(base),
    result = isString(baseValue) ? trimTrailingSemicolon(baseValue) : '';

  if (Object.keys(tokens).length > 0) {
    const styleMap = new Map<string, string>();
    parseStyleAttr(styleMap, result);
    resolveStyles(styleMap, tokens);
    result = '';
    for (const [name, value] of styleMap) result += `${name}: ${value};`;
  }

  result = result.trim();
  return result.length ? ` style="${escape(result, true)}"` : result;
}

const styleSplitRE = /\s*:\s*/;
const stylesDelimeterRE = /\s*;\s*/;
export function parseStyleAttr(tokens: Map<string, string>, attrValue: string) {
  const styles = attrValue.trim().split(stylesDelimeterRE);
  for (let i = 0; i < styles.length; i++) {
    if (styles[i] === '') continue;
    const [name, value] = styles[i].split(styleSplitRE);
    tokens.set(name, value);
  }
}

function resolveStyles(tokens: Map<string, string>, styles: Record<string, unknown>) {
  for (const name of Object.keys(styles)) {
    const value = unwrapDeep(styles[name]);
    if (!value && value !== 0) {
      tokens.delete(name);
    } else {
      tokens.set(name, value + '');
    }
  }
}

const propNameRE = /[A-Z]/;
export function $$_merge_spreads(spreads: Record<string, unknown>[]) {
  let attributes = new Map<string, string>(),
    classList = new Set<string>(),
    styles = new Map<string, string>();

  for (let i = 0; i < spreads.length; i++) {
    // $$class and $$style are dynamics
    const { $$class, $$style, class: classBase, style: styleBase, ...attrs } = spreads[i];

    const attrNames = Object.keys(attrs);
    for (let j = 0; j < attrNames.length; j++) {
      const attrName = attrNames[j];
      if (!propNameRE.test(attrName)) {
        const attrValue = resolveAtrr(attrs[attrName]);
        if (isString(attrValue)) {
          attributes.set(attrName, attrValue);
        } else {
          attributes.delete(attrName);
        }
      }
    }

    if ('class' in spreads[i]) {
      const base = unwrapDeep(classBase);
      if (isNull(base) || base === false) {
        classList.clear();
      } else if (isString(base) && base.length) {
        parseClassAttr(classList, base + '');
      }
    }

    if ($$class) resolveClasses(classList, $$class as Record<string, unknown>);

    if ('style' in spreads[i]) {
      const base = unwrapDeep(styleBase);
      if (isNull(base) || base === false) {
        styles.clear();
      } else if (isString(base) && base.length) {
        parseStyleAttr(styles, base);
      }
    }

    if ($$style) resolveStyles(styles, $$style as Record<string, unknown>);
  }

  return { attributes, classList, styles };
}

/** @internal */
export function $$_spread(spreads: Record<string, unknown>[]) {
  const { attributes, classList, styles } = $$_merge_spreads(spreads);

  let result = '';

  if (classList.size > 0) {
    let _class = Array.from(classList).join(' ');
    result += `class="${escape(_class.trim(), true)}"`;
  }

  for (const [name, value] of attributes) {
    result += ` ${name}="${escape(value, true)}"`;
  }

  if (styles.size > 0) {
    let _styles = '';
    for (const [name, value] of styles) _styles += `${name}: ${value};`;
    result += ` style="${escape(_styles.trim(), true)}"`;
  }

  return result.trim();
}

export function $$_inject_html(value: unknown): any {
  return injectHTML(resolve(value));
}

export { $$_merge_props, $$_create_component } from '../dom';
