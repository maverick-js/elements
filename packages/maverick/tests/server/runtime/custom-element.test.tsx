import { CustomElement } from 'maverick.js';

import {
  createElementInstance,
  createServerElement,
  css,
  defineCustomElement,
} from 'maverick.js/element';
import { renderToString } from 'maverick.js/ssr';

it('should render', () => {
  const element = defineCustomElement({
    tagName: 'mk-foo',
    setup: () => () => <div class="foo">Test</div>,
  });

  const result = renderToString(() => <CustomElement $element={element} />).code;

  expect(result).toMatchInlineSnapshot(
    '"<!$><mk-foo mk-h=\\"\\" mk-d=\\"\\"><shadow-root><!$><div class=\\"foo\\">Test</div></shadow-root></mk-foo>"',
  );
});

it('should render with children', () => {
  const element = defineCustomElement({
    tagName: 'mk-foo',
    setup: () => () => <div class="foo">Test</div>,
  });

  const result = renderToString(() => (
    <CustomElement $element={element}>
      <div>Child</div>
    </CustomElement>
  )).code;

  expect(result).toMatchInlineSnapshot(
    '"<!$><mk-foo mk-h=\\"\\" mk-d=\\"\\"><shadow-root><!$><div class=\\"foo\\">Test</div></shadow-root><!$><div>Child</div></mk-foo>"',
  );
});

it('should render with shadow dom', () => {
  const element = defineCustomElement({
    tagName: 'mk-foo',
    shadowRoot: true,
    setup: () => () => <div class="foo">Test</div>,
  });

  const result = renderToString(() => (
    <CustomElement $element={element}>
      <div>Light A</div>
      <div>Light B</div>
      <div>Light C</div>
    </CustomElement>
  )).code;

  expect(result).toMatchInlineSnapshot(
    '"<!$><mk-foo mk-h=\\"\\" mk-d=\\"\\"><template shadowroot=\\"open\\"><!$><div class=\\"foo\\">Test</div></template><!$><div>Light A</div><!$><div>Light B</div><!$><div>Light C</div></mk-foo>"',
  );
});

it('should render adopted css styles in shadow root template', () => {
  const Button = defineCustomElement({
    tagName: `mk-button-7`,
    shadowRoot: true,
    css: [
      css`
        div {
          display: inline-block;
        }
      `,
      css`
        .container {
          color: blue;
          order: ${4};
        }
      `,
    ],
  });

  const instance = createElementInstance(Button);
  const element = new (createServerElement(Button))();
  element.attachComponent(instance);

  // Fragile snapshot can't inline.
  expect(element.render()).toMatchSnapshot();
});

it('should render with attributes', () => {
  const element = defineCustomElement({ tagName: 'mk-foo' });

  const result = renderToString(() => (
    <CustomElement
      foo={10}
      bar={'boo'}
      $class:foo={true}
      $class:bar={true}
      $style:display={'none'}
      $cssvar:baz={10}
      $element={element}
    />
  )).code;

  expect(result).toMatchInlineSnapshot(
    '"<!$><mk-foo foo=\\"10\\" bar=\\"boo\\" mk-h=\\"\\" mk-d=\\"\\" class=\\"foo bar\\" style=\\"display: none;--baz: 10;\\"></mk-foo>"',
  );
});

it('should forward props', () => {
  const element = defineCustomElement({
    tagName: 'mk-foo',
    // @ts-expect-error
    props: {
      foo: { initial: 10 },
    },
    setup:
      ({ props }) =>
      () =>
        <div>{props.foo}</div>,
  });

  const result = renderToString(() => <CustomElement $prop:foo={100} $element={element} />).code;

  expect(result).toMatchInlineSnapshot(
    '"<!$><mk-foo mk-h=\\"\\" mk-d=\\"\\"><shadow-root><!$><div><!$>100</div></shadow-root></mk-foo>"',
  );
});

it('should set inner html', () => {
  const element = defineCustomElement({
    tagName: 'mk-foo-10',
    setup: () => () => <div>Test</div>,
  });

  const result = renderToString(() => (
    <CustomElement $prop:innerHTML={`<div>INNER HTML</div>`} $element={element}>
      <div>Foo</div>
    </CustomElement>
  )).code;

  expect(result).toMatchInlineSnapshot(
    '"<!$><mk-foo-10 mk-h=\\"\\" mk-d=\\"\\"><div>INNER HTML</div></mk-foo-10>"',
  );
});

it('should render `setAttributes`', () => {
  const Foo = defineCustomElement({
    tagName: `mk-foo-1`,
    setup({ host }) {
      host.setAttributes({
        foo: () => 10,
        bar: 'none',
        baz: null,
        bux: false,
      });
    },
  });

  const instance = createElementInstance(Foo);
  const element = new (createServerElement(Foo))();
  element.attachComponent(instance);

  expect(element.attributes.tokens).toMatchInlineSnapshot(`
    Map {
      "mk-h" => "",
      "mk-d" => "",
      "foo" => "10",
      "bar" => "none",
    }
  `);
});

it('should render `setStyles`', () => {
  const Foo = defineCustomElement({
    tagName: `mk-foo-2`,
    setup({ host }) {
      host.setStyles({
        flex: '1',
        'flex-basis': null,
        'align-self': false,
        'z-index': () => 10,
      });
    },
  });

  const instance = createElementInstance(Foo);
  const element = new (createServerElement(Foo))();
  element.attachComponent(instance);

  expect(element.style.tokens).toMatchInlineSnapshot(`
    Map {
      "flex" => "1",
      "z-index" => "10",
    }
  `);
});

it('should render `setCSSVars`', () => {
  const Foo = defineCustomElement({
    tagName: `mk-foo-3`,
    setup({ host }) {
      host.setCSSVars({
        '--foo': () => 10,
        '--bar': 'none',
        '--baz': false,
        '--bax': null,
      });
    },
  });

  const instance = createElementInstance(Foo);
  const element = new (createServerElement(Foo))();
  element.attachComponent(instance);

  expect(element.style.tokens).toMatchInlineSnapshot(`
    Map {
      "--foo" => "10",
      "--bar" => "none",
    }
  `);
});
