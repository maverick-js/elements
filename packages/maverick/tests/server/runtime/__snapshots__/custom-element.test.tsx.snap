// Vitest Snapshot v1

exports[`should render adopted css styles in shadow root template 1`] = `
"<template shadowroot=\\"open\\"><style>
        div {
          display: inline-block;
        }
      
        .container {
          color: blue;
          order: 4;
        }
      </style></template>"
`;
